### ISSUE TITLE
CosineAnnealingWarmRestarts documentation poor and not appearing

### ISSUE THREAD

**Issue by jcreinhold** [pytorch/pytorch#20028]
## 📚 Documentation

The documentation for the newly introduced `CosineAnnealingWarmRestarts` learning rate scheduler (#17226) does not appear on the website (see [here](https://pytorch.org/docs/stable/optim.html); the location where it should be).

Furthermore, looking at the [source code of the function](https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L699), I am very confused about how to use the scheduler.

Specifically, if we look at the documentation of `.step()`, we first see a clear typo, where `CosineAnnealingWarmRestarts` is called `SGDR`. But the provided example is not sufficient to use as a reference. Where did the value 26 come from and why, and what does it have to do with 10 iterations? Also, [the paper](https://arxiv.org/abs/1608.03983) on which this scheduler is based states that the LR should be updated every _batch_, not epoch. I believe the statement in the documentation `SGDR.step(0.1), SGDR.step(0.2)` is referring to this, but I am lost in the example.

If I try to ignore the docstring and look at the source code, my best guess is that the correct message for the docstring in `.step` should be something like:

```python
"""Step should be called after every time a batch is processed in addition to after each epoch,
   like in the following example:

         >>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
         >>> for epoch in range(N):
         >>>     scheduler.step(epoch)
         >>>     for inputs, labels in trainloader:
         >>>         optimizer.zero_grad()
         >>>         outputs = net(inputs)
         >>>         loss = criterion(outputs, labels)
         >>>         loss.backward()
         >>>         optimizer.step()
         >>>         scheduler.step()
"""
```

Is this correct? Or have I misinterpreted how to apply this function?

Thanks for the great work and please let me know if I can help in some way.

**Comment by Kirayue** [comment]
Hi, @jcreinhold
I also wondered why it did not appear. Sorry for the poor doc after deciding to change name form SGDR to CosineAnnealingWarmRestarts.
About the usage `LR should be updated every batch, not epoch`, `step()`  without argument is set to epoch update, you can give an argument for a batch update. For example, if you have 15 training examples and batch size is 4, so there would be 4 iterations in an epoch. You could call step(0 + 0 / 4) for the epoch 0 and first batch update, step(0 + 1 / 4) for the epoch 0 and second batch update.
So the code might be modified to 
```
"""Step should be called after every time a batch is processed in addition to after each epoch,
   like in the following example:

         >>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
         >>> iters = len(trainloader)
         >>> for epoch in range(N):
         >>>     scheduler.step(epoch)
         >>>     for iter, (inputs, labels) in enumerate(trainloader):
         >>>         optimizer.zero_grad()
         >>>         outputs = net(inputs)
         >>>         loss = criterion(outputs, labels)
         >>>         loss.backward()
         >>>         optimizer.step(epoch + i / total_iters)
         >>>         scheduler.step()
"""
```

**Comment by jcreinhold** [comment]
@Kirayue I know this is not relevant to the current discussion, but—since you implemented and contributed the method to pytorch—what are your thoughts on adding in "normalized weight decay" as per Appendix B.1 in [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101), which is, for all intents and purposes, the follow-up paper for SGDR.

They say it is a preliminary result, but "_some_ normalization [on weight decay] can substantially improve results." Any thoughts?

Thanks for the great work and for contributing.

**Comment by Kirayue** [comment]
@jcreinhold, I will take it into consideration and implement it.

Thank you for the suggestions.

**Comment by Chillee** [comment]
Not all the aspects of this issue have been resolve. For example, it still doesn't show up here: https://pytorch.org/docs/master/optim.html#how-to-adjust-learning-rate.

**Comment by calincru** [comment]
The documentation is still very confusing.  I'm still not sure how it differs from `torch.optim.lr_scheduler.CosineAnnealingLR`.  Is it just the flexibility provided by `T_mult`?  I feel there should be at least some note on this.

Also, slightly related: what I'm interested in is a phase-shifted version of cosine annealing where the training starts with the low learning rates, increases, and then decreases again, and so on (essentially I want the same formula but with `torch.sin` instead of `torch.cos`).  Is it possible to achieve this with one of the existing `CosineAnnealing*` classes?

**Comment by Kirayue** [comment]
Hi, @calincru
The `CosineAnneling` and `Restarts` are both mentioned in [SGDR: Stochastic Gradient Descent with Warm Restarts](https://arxiv.org/abs/1608.03983), so the document looks similar. But the former is more like to [Cyclical Learning Rates](https://arxiv.org/pdf/1506.01186.pdf). 

BTW, in my opinion, `CosineAnnealing` and `ConsineAnnealingWarmRestarts` are kinds of  `Cyclical LR` (although `CosineAnnealingWarmRestarts` has `T_mult` which makes it not periodic). 

I plot the two lr schedulers on tensorboard, there is a restart point in `CosineAnnealingWarmRestarts` right after reaching the lowest point.
#### CosineAnnealingLR
![image](https://user-images.githubusercontent.com/13309212/63071590-7acd2900-bf52-11e9-909e-85f67107ee33.png)
#### CosineAnnealingWarmRestarts
![image](https://user-images.githubusercontent.com/13309212/63071620-9801f780-bf52-11e9-9b68-f84be8e48190.png)

My torch version is 1.1.0.

I think there is no phase or related argument to achieve `sin` but you can modify `CosineAnnealingWarmRestart` code to 
```
def get_lr(self):
    return [self.eta_min + (base_lr - self.eta_min) * (1 + math.cos(math.pi * self.T_cur / self.T_i + math.pi)) / 2 for base_lr in self.base_lrs]
```

**Comment by calincru** [comment]
@Kirayue Thanks for clarifications!

**Comment by peter850706** [comment]
Hi, @Kirayue
I read your comments above that mention your torch version is 1.1.0. However, [this warning](https://pytorch.org/docs/master/optim.html#how-to-adjust-learning-rate) indicates that the `scheduler.step()` should be called after `optimizer.step()`.
So I wonder if the example of [CosineAnnealingWarmRestart](https://pytorch.org/docs/master/optim.html#torch.optim.lr_scheduler.CosineAnnealingWarmRestarts) should be modified as the following:
```
Example:
    >>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
    >>> iters = len(dataloader)
    >>> for epoch in range(20):
    >>>     for i, sample in enumerate(dataloader):
    >>>         inputs, labels = sample['inputs'], sample['labels']
    >>>         optimizer.zero_grad()
    >>>         outputs = net(inputs)
    >>>         loss = criterion(outputs, labels)
    >>>         loss.backward()
    >>>         optimizer.step()
    >>>         scheduler.step(epoch + i / iters)
````

**Comment by AntiLibrary5** [comment]
Docs still hasn't been fixed:

https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html
